### PR TITLE
Handle live fiber snapshot signals on the JVM

### PR DESCRIPTION
**Pull request by vasilmkd**
This PR registers the signals for handling live fiber snapshots on the JVM. It also sets up an integration test.

The bulk of the changes are in test code. All affected runtimes are tested.

### CODE REVIEW

**Comment by vasilmkd** [comment]
```
load: 16.37  cmd: java 71964 running 421.65u 1.63s
[error] Live Fiber Snapshot
[error] Worker Thread #4 (Thread[io-compute-4,5,main])
[error] Active fiber: cats.effect.IOFiber@4b22e35d RUNNABLE
[error] Enqueued fibers:
[error] cats.effect.IOFiber@1c5c0119 YIELDING
[error] Worker Thread #0 (Thread[io-compute-0,5,main])
[error] Active fiber: cats.effect.IOFiber@2f886aee RUNNABLE
[error] Enqueued fibers:
[error] cats.effect.IOFiber@11dad897 YIELDING
[error] cats.effect.IOFiber@2f886aee YIELDING
[error] Worker Thread #2 (Thread[io-compute-2,5,main])
[error] Active fiber: cats.effect.IOFiber@4dbee9fa RUNNABLE
[error] Enqueued fibers:
[error] Worker Thread #7 (Thread[io-compute-7,5,main])
[error] Active fiber: cats.effect.IOFiber@5843d96f RUNNABLE
[error] Enqueued fibers:
[error] cats.effect.IOFiber@38cb953e YIELDING
[error] cats.effect.IOFiber@199e1b5a YIELDING
[error] cats.effect.IOFiber@657dc691 YIELDING
[error] cats.effect.IOFiber@3b562d06 YIELDING
[error] cats.effect.IOFiber@7d1803a2 YIELDING
[error] cats.effect.IOFiber@5843d96f YIELDING
[error] cats.effect.IOFiber@5c978fe6 YIELDING
[error] cats.effect.IOFiber@6e1cc966 YIELDING
[error] cats.effect.IOFiber@66c77bd4 YIELDING
[error] cats.effect.IOFiber@330985f8 YIELDING
[error] Worker Thread #6 (Thread[io-compute-6,5,main])
[error] Active fiber: cats.effect.IOFiber@3497c8d1 RUNNABLE
[error] Enqueued fibers:
[error] cats.effect.IOFiber@3f0e0821 YIELDING
[error] cats.effect.IOFiber@30e761b1 YIELDING
[error] cats.effect.IOFiber@20f8c6ad YIELDING
[error] cats.effect.IOFiber@233c5283 YIELDING
[error] cats.effect.IOFiber@72d5556d YIELDING
[error] cats.effect.IOFiber@1be67036 YIELDING
[error] cats.effect.IOFiber@61112846 YIELDING
[error] cats.effect.IOFiber@6c3c0025 YIELDING
[error] cats.effect.IOFiber@3497c8d1 YIELDING
[error] cats.effect.IOFiber@b73e112 YIELDING
[error] cats.effect.IOFiber@585c21e4 YIELDING
[error] cats.effect.IOFiber@349a45e6 YIELDING
[error] Worker Thread #3 (Thread[io-compute-3,5,main])
[error] Active fiber: cats.effect.IOFiber@210d25b9 RUNNABLE
[error] Enqueued fibers:
[error] Worker Thread #5 (Thread[io-compute-5,5,main])
[error] Active fiber: cats.effect.IOFiber@5e4714e1 RUNNABLE
[error] Enqueued fibers:
[error] Worker Thread #1 (Thread[io-compute-1,5,main])
[error] Active fiber: cats.effect.IOFiber@199e71de RUNNABLE
[error] Enqueued fibers:
[error] Fibers enqueued on the external queue of the Work Stealing Runtime:
[error] Asynchronously suspended fibers:
[error] cats.effect.IOFiber@46e1db3b WAITING
[error] cats.effect.IOFiber@5c456007 WAITING
[error] cats.effect.IOFiber@609ff839 WAITING
[error] cats.effect.IOFiber@315f9299 WAITING
[error] cats.effect.IOFiber@25c1a6f9 WAITING
[error] cats.effect.IOFiber@7f4a4b81 WAITING
[error] cats.effect.IOFiber@6e3f5e6b WAITING
[error] cats.effect.IOFiber@717e2e4d WAITING
[error] cats.effect.IOFiber@3153deb8 WAITING
[error] Fibers executing on foreign Execution Contexts:
```

**Comment by vasilmkd** [comment]
Now to figure out why the output doesn't show up in tests...

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ jobs:
     name: Build and Test
     strategy:
       matrix:
-        os: [ubuntu-latest, windows-latest]
+        os: [ubuntu-latest, windows-latest, macos-latest]
```
Comment (armanbilge) [inline .github/workflows/ci.yml:25]: The matrix grows again 😆

File: core/jvm/src/main/scala/cats/effect/IOApp.scala [hunk 2]
```diff
@@ -231,6 +234,38 @@ trait IOApp {
...
+            case _ =>
+              // MacOS and BSD can handle both USR1 (except JDK 8) and INFO (nice CTRL+T experience)
+              List("USR1", "INFO")
+          }
+        }
+
+      liveFiberSnapshotSignal.map(new Signal(_)).foreach { signal =>
+        Signal.handle(
+          signal,
+          _ => runtime.fiberMonitor.liveFiberSnapshot().foreach(System.err.println(_)))
```
Comment (armanbilge) [inline core/jvm/src/main/scala/cats/effect/IOApp.scala]: Not sure about this (or if it's important anyway) but would concat-ing the lines and then `println`ing once have a better chance of being atomic than `println`ing each line seperately?

**Review by armanbilge — DISMISSED** [review]
Lot of respect for shaving the testing yak here. We should be able to reuse it for JS, still annoying but will be much less so.

File: core/jvm/src/main/scala/cats/effect/IOApp.scala (hunk 2 above)
Comment (vasilmkd) [reply core/jvm/src/main/scala/cats/effect/IOApp.scala]: That's already the case. `def liveFiberSnapshot(): Option[String]`. Foreach is just a generic Scala collections method that works on the `Some` case of the `Option`.

Comment (armanbilge) [reply core/jvm/src/main/scala/cats/effect/IOApp.scala]: Ah, my bad, sorry! I thought it was a List. 👍

File: tests/jvm/src/test/scala/cats/effect/IOAppSpec.scala [hunk 3]
```diff
@@ -127,14 +142,50 @@ class IOAppSpec extends Specification {
...
+    val jpsProcess =
+      Process(s"$JavaHome/bin/jps", List.empty).run(BasicIO(false, jpsStdoutBuffer, None))
+    jpsProcess.exitValue()
+
+    val output = jpsStdoutBuffer.toString
+    Source.fromString(output).getLines().find(_.contains(mainName)).map(_.split(" ")(0).toInt)
+  }
+
+  private def sendSignal(pid: Int): Unit = {
+    Runtime.getRuntime().exec(s"kill -USR1 $pid")
```
Comment (djspiewak) [inline tests/jvm/src/test/scala/cats/effect/IOAppSpec.scala:179]: Shouldn't this *not* work on macOS JDK 8?

**Review by djspiewak — COMMENTED** [review]
Mad respect on the testing yak. This is a nice approach.

**Comment by djspiewak** [comment]
Interestingly, `INFO` doesn't work on macOS on the following VM version:

```
java version "11.0.1" 2018-10-16 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.1+13-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.1+13-LTS, mixed mode)
```

**Comment by vasilmkd** [comment]
I used the following resources:
[JDK 17](https://docs.oracle.com/en/java/javase/17/troubleshoot/handle-signals-and-exceptions.html#GUID-57C048F6-0D4B-43BD-B27C-06A613435360), [JDK 11](https://docs.oracle.com/en/java/javase/11/troubleshoot/handle-signals-and-exceptions.html#GUID-57C048F6-0D4B-43BD-B27C-06A613435360), [JDK 8](https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/signals006.html).

On JDK8, `SIGUSR1` used to be reserved, hence the filtering in the code, but this is not the case since 9+. There is no mention of `INFO` anywhere.

> java version "11.0.1" 2018-10-16 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.1+13-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.1+13-LTS, mixed mode)

This is a 3 years out-of-date version of the JDK. We're technically not obliged to support it. Also nowhere in the spec is `INFO` mentioned as reserved.

Can you post the exception?

Fwiw, I was against signals in the beginning. 🤷🏻

**Comment by vasilmkd** [comment]
Discussion about using unsupported classes and the Java module system support in Scala. Maybe we should think twice before relying on `sun.misc.Signal`.

https://discord.com/channels/632277896739946517/839263556754472990/909397558395306034

**Review by djspiewak — APPROVED** [review]
We definitely need to adjust the formatting before release, but that's what #2542 does

**Pull request merged**